File: sdk/python/alp_sdk/event_mesh.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import List, Dict, Any, Optional, Callable, Set

class MeshEvent:
    __test__ = False

    def __init__(
        self,
        event_id: str,
        topic: str,
        sender_agent: str,
        payload: str,
        event_type: str = "state_change",
        timestamp: Optional[str] = None,
    ):
        self.id = event_id
        self.topic = topic
        self.sender_agent = sender_agent
        self.payload = payload
        self.event_type = event_type
        self.timestamp = timestamp or datetime.now(timezone.utc).isoformat()

class EventMeshConfig:
    __test__ = False

    def __init__(
        self,
        event_id: str,
        topic: str,
        sender_agent: str,
        payload: str,
        event_type: str = "state_change",
        timestamp: Optional[str] = None,
        description: Optional[str] = None,
    ):
        self.id = event_id
        self.topic = topic
        self.sender_agent = sender_agent
        self.payload = payload
        self.event_type = event_type
        self.timestamp = timestamp or datetime.now(timezone.utc).isoformat()
        self.description = description

class EventMeshEngine:
    def __init__(self):
        self.subscriptions: Dict[str, Set[Callable[[MeshEvent], None]]] = {}
        self.event_buffer: List[MeshEvent] = []

    def subscribe(self, topic: str, handler: Callable[[MeshEvent], None]) -> Callable[[], None]:
        subs = self.subscriptions.get(topic, set())
        subs.add(handler)
        self.subscriptions[topic] = subs

        def unsubscribe():
            if topic in self.subscriptions and handler in self.subscriptions[topic]:
                self.subscriptions[topic].remove(handler)

        return unsubscribe

    def publish(
        self,
        event_id: str,
        topic: str,
        sender_agent: str,
        payload: str,
        event_type: str = "state_change",
    ) -> MeshEvent:
        event = MeshEvent(
            event_id=event_id,
            topic=topic,
            sender_agent=sender_agent,
            payload=payload,
            event_type=event_type,
        )
        self.event_buffer.append(event)

        topic_subs = self.subscriptions.get(topic, set())
        for handler in list(topic_subs):
            handler(event)

        wildcard_subs = self.subscriptions.get("*", set())
        for handler in list(wildcard_subs):
            handler(event)

        return event

    def get_event_history(self, topic: Optional[str] = None) -> List[MeshEvent]:
        if not topic or topic == "*":
            return list(self.event_buffer)
        return [e for e in self.event_buffer if e.topic == topic]

    def clear_buffer(self) -> None:
        self.event_buffer.clear()

File: parser/src/event-mesh.ts
export type EventType = 'state_change' | 'task_update' | 'agent_broadcast' | 'alert';

export interface MeshEvent {
  id: string;
  topic: string;
  senderAgent: string;
  payload: string;
  eventType: EventType;
  timestamp: string;
}

export type EventSubscriber = (event: MeshEvent) => void;

export class EventMeshEngine {
  private subscriptions: Map<string, Set<EventSubscriber>> = new Map();
  private eventBuffer: MeshEvent[] = [];

  public subscribe(topic: string, handler: EventSubscriber): () => void {
    const subs = this.subscriptions.get(topic) || new Set();
    subs.add(handler);
    this.subscriptions.set(topic, subs);

    return () => {
      subs.delete(handler);
    };
  }

  public publish(
    id: string,
    topic: string,
    senderAgent: string,
    payload: string,
    eventType: EventType = 'state_change'
  ): MeshEvent {
    const event: MeshEvent = {
      id,
      topic,
      senderAgent,
      payload,
      eventType,
      timestamp: new Date().toISOString(),
    };

    this.eventBuffer.push(event);

    const topicSubscribers = this.subscriptions.get(topic);
    if (topicSubscribers) {
      topicSubscribers.forEach(handler => handler(event));
    }

    const wildcardSubscribers = this.subscriptions.get('*');
    if (wildcardSubscribers) {
      wildcardSubscribers.forEach(handler => handler(event));
    }

    return event;
  }

  public getEventHistory(topic?: string): MeshEvent[] {
    if (!topic || topic === '*') {
      return [...this.eventBuffer];
    }
    return this.eventBuffer.filter(e => e.topic === topic);
  }

  public clearBuffer(): void {
    this.eventBuffer = [];
  }
}
